pkg/controller/vitessbackupschedule/vitessbackupschedule_controller.go

```diff
@@ -0,0 +1,657 @@
+// It also returns at what time was the last job created, which is needed to update VitessBackupSchedule's status,
+// and plan future jobs.
+func (r *ReconcileVitessBackupsSchedule) getJobsList(ctx context.Context, req ctrl.Request, vbscName string) (jobsList, *time.Time, error) {
+	var existingJobs kbatch.JobList
+
+	err := r.client.List(ctx, &existingJobs, client.InNamespace(req.Namespace), client.MatchingLabels{planetscalev2.BackupScheduleLabel: vbscName})
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.WithError(err).Error("unable to list Jobs in cluster")
+		return jobsList{}, nil, err
+	}
+
+	var jobs jobsList
+
+	var mostRecentTime *time.Time
+
+	for i, job := range existingJobs.Items {
+		_, finishedType := isJobFinished(&job)
+		switch finishedType {
+		case kbatch.JobFailed:
+			jobs.failed = append(jobs.failed, &existingJobs.Items[i])
+		case kbatch.JobComplete:
+			jobs.successful = append(jobs.successful, &existingJobs.Items[i])
+		default:
+			// Either: Suspended, FailureTarget or simply ongoing
+			jobs.active = append(jobs.active, &existingJobs.Items[i])
+		}
+
+		scheduledTimeForJob, err := getScheduledTimeForJob(&job)
+		if err != nil {
+			log.WithError(err).Errorf("unable to parse schedule time for existing job, found: %s", job.Annotations[scheduledTimeAnnotation])
+			continue
+		}
+		if scheduledTimeForJob != nil && (mostRecentTime == nil || mostRecentTime.Before(*scheduledTimeForJob)) {
+			mostRecentTime = scheduledTimeForJob
+		}
+	}
+	return jobs, mostRecentTime, nil
+}
+
+// cleanupJobsWithLimit removes all Job objects from the cluster ordered by oldest to newest and
+// respecting the given limit, keeping minimum "limit" jobs in the cluster.
+func (r *ReconcileVitessBackupsSchedule) cleanupJobsWithLimit(ctx context.Context, jobs []*kbatch.Job, limit int32) {
+	if limit == -1 {
+		return
+	}
+
+	sort.SliceStable(jobs, func(i, j int) bool {
+		if jobs[i].Status.StartTime == nil {
+			return jobs[j].Status.StartTime != nil
+		}
+		return jobs[i].Status.StartTime.Before(jobs[j].Status.StartTime)
+	})
+
+	for i, job := range jobs {
+		if int32(i) >= int32(len(jobs))-limit {
+			break
+		}
+		if err := r.client.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground)); (err) != nil {
+			log.WithError(err).Errorf("unable to delete old job: %s", job.Name)
+		} else {
+			log.Infof("deleted old job: %s", job.Name)
+		}
+	}
+}
+
+func (r *ReconcileVitessBackupsSchedule) removeTimeoutJobs(ctx context.Context, jobs []*kbatch.Job, timeout int32) error {
+	if timeout == -1 {
+		return nil
+	}
+	for _, job := range jobs {
+		jobStartTime, err := getScheduledTimeForJob(job)
+		if err != nil {
+			return err
+		}
+		if jobStartTime.Add(time.Minute * time.Duration(timeout)).Before(time.Now()) {
+			if err := r.client.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground)); (err) != nil {
```

Comment:
seems like a good thing to have a metric for

Reply:
fixed via 46b6967